Add --force to upgrade and rollback to delete/create when conflicts

peay opened the pull request:
Closes #1958 

I have added `--force` to upgrade, and rollback. This is consistent with `--recreate-pods` not being available in `install`, and with the semantics of `install` vs these two I believe.

A couple remarks:
- There are no timeout and grace period parameters -- the per-resource defaults are used.
- There are no multiple retries either, unlike in `kubectl apply --force`.
- I have not added tests yet - suggestions on how to tackle this best would be helpful.
- I am not sure if there are situations where we succeed in deleting the resource, but not in re-creating. Most of the validations on the new structure have already been performed at this point.

I have split the modifications into two commits with and without generated code to make review easier.

cc @thomastaylor312

thomastaylor312 commented:
@peay Would you like us to take a look at this now or are you still working on a few things?

peay commented:
I am done -- Can you please have a look? Thanks!

thomastaylor312 commented:
@technosophos This makes a change to the `Update` method in the release server. Are we ok with introducing a breaking change (albeit small) like this in 2.4?

@peay I'll be reviewing this in the evening or tomorrow

thomastaylor312 commented:
So in answer to your testing question, there isn't a clean way to test these methods that use a client set. We really want to start doing integration tests to cover those

Also, I was re-reviewing #1958 and was thinking that it may be a good idea to have a `--grace-period` flag included because the user may want to override the default value. @adamreese what do you think?

thomastaylor312 reviewed (changes requested):
I am excited to see this feature land. I found a few edge cases while testing.

If I make a change to an immutable field and then try to upgrade without `--force`, it errors out as expected. However, when I add the flag after doing that, it won't upgrade because it stored the manifest for the failed upgrade and now thinks there are no changes.

The second problem is how force interacts with the `--wait` flag. If I use both flags, I get this error
```
Error: UPGRADE FAILED: Object 'Kind' is missing in 'null
'
```

I didn't debug it fully, but I am betting that the `--wait` flag can't find the old resource that was deleted by `--force`. It still deletes and recreates it properly, but it breaks the wait functionality.

Also, as a note, one of your files is not gofmt'd and the spelling error noted below are what is causing the test failures

Anyway, I know those are some big things to address but this is honestly a great first pass. Let us know if you have any questions

peay commented on pkg/proto/hapi/chart/chart.pb.go:
```diff
@@ -101,7 +101,7 @@ func init() { proto.RegisterFile("hapi/chart/chart.proto", fileDescriptor0) }
 var fileDescriptor0 = []byte{
 	// 242 bytes of a gzipped FileDescriptorProto
-	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x09, 0x6e, 0x88, 0x02, 0xff, 0x6c, 0x90, 0xb1, 0x4e, 0xc3, 0x30,
+	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0xff, 0x6c, 0x90, 0xb1, 0x4e, 0xc3, 0x30,
```
Indeed:

```
protoc --version                                                                                                                                                                                                                                                                                                                                                  libprotoc 3.2.0
```

I'll rebase, thanks.

peay committed:
Add --force to upgrade and rollback

peay committed:
Add --force to upgrade and rollback (generated changes)

peay commented:
@thomastaylor312 thanks a lot for the comments. I just pushed some modifications:

- Fixed the typo, tests pass now.
- Fixed the issue with `--wait`. When we recreate, we should not `target.Refresh`: this is not useful anyway as we recreated the resource based on `target`.

Regarding the fact that a manifest that we fail to apply stills gets stored, this seems like an issue unrelated to the addition `--force`, right? Should this be tackled in a separate issue/PR?

This could be via `--refresh` argument if we want to keep it optional, but maybe you want to make it the default to refresh resources before doing an update.

thomastaylor312 commented:
@peay With this change, will `--wait` still wait for a resource that was newly created with `--force`?

peay commented:
@thomastaylor312 I think so, why not? Now, re-creates behaves as the code for the original creation, which does not call `Refresh` as the description in `target` is what has been used to create the resource anyway.

thomastaylor312 reviewed (approved):
Ok, this looks good to me! Everything is tested and good to go. I am going to wait to merge until I hear from one of the other maintainers

adamreese commented:
Hi @peay!  How can this be tested manually?

adamreese commented on pkg/tiller/release_server.go:
```diff
@@ -881,7 +882,8 @@ func (s *ReleaseServer) performRelease(r *release.Release, req *services.Install
 		r.Version = old.Version + 1
 
-		if err := s.performKubeUpdate(old, r, false, req.Timeout, req.Wait); err != nil {
+		// TODO: why is req.Recreate not used here? Not using force either for consistency
```
Probably just never added the cli flag on `helm install`

adamreese commented on pkg/tiller/release_server.go:
```diff
@@ -525,11 +525,12 @@ func (s *ReleaseServer) performRollback(currentRelease, targetRelease *release.R
 
-func (s *ReleaseServer) performKubeUpdate(currentRelease, targetRelease *release.Release, recreate bool, timeout int64, shouldWait bool) error {
+func (s *ReleaseServer) performKubeUpdate(currentRelease, targetRelease *release.Release, force bool, recreate bool, timeout int64, shouldWait bool) error {
+	log.Printf("perform: force: %v", force)
```
Not sure how much value this adds.  Can it be removed or moved somewhere more pertinent

peay commented:
@adamreese good catch, forgot to remove this. I have been testing manually and using variants of this for lack of something better:

```

% cat test-svc-tmpl/templates/test-svc.yaml

apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  ports:
  - name: cluster
    port: 7000
    targetPort: 7000
  selector:
    name: test
  type: ClusterIP

% cat test-force.sh

#!/bin/bash

set -e

DUMMY_IP=10.233.39.3

function install() {
    bin/helm delete --purge test-svc

    rm -rf test-svc
    cp -R test-svc-tmpl test-svc
    bin/helm upgrade --install test-svc test-svc/
}

function test_without_force() {
    install

    echo "  clusterIP: $DUMMY_IP" >> test-svc/templates/test-svc.yaml
    ! (bin/helm upgrade --install test-svc test-svc/)

    sleep 5
    kubectl get svc -oyaml | grep -v "$DUMMY_IP"
}

function test_force() {
    install

    echo "  clusterIP: $DUMMY_IP" >> test-svc/templates/test-svc.yaml
    bin/helm upgrade --install test-svc test-svc/ --force

    sleep 5
    kubectl get svc -oyaml | grep "$DUMMY_IP"
}

function test_force_wait() {
    install

    echo "  clusterIP: $DUMMY_IP" >> test-svc/templates/test-svc.yaml
    bin/helm upgrade --install test-svc test-svc/ --force --wait

    kubectl get svc -oyaml | grep "$DUMMY_IP"
}

set -x
test_without_force
test_force
test_force_wait
set +x

echo "OK"
```

technosophos commented:
Can we get a rebase on this one in time for 2.5.0?

peay commented:
I'll update the PR before the end of the week, does that work?

peay commented:
@technosophos @thomastaylor312 rebased. There were a couple of changes due to API changes with the new `ReleaseModule`, so please have a look.

@thomastaylor312 there are a couple generated changes in Protobuf outside of what I modified. I see that your PR for `protoc 3.2.0` was merged, and this is also what I use, not sure why it generated a diff on all those files. Maybe someone forgot to upgrade?

I've again separated commits to make review easier.

peay commented:
Also, CI seems to fail, but `make bootstrap build protoc docs verify-docs` succeeds and leaves HEAD untouched on my local copy, not sure why.

thomastaylor312 commented:
@peay Looks like the docs need to be updated. This is the error from circleci: `helm docs are out of date. Please run "make docs"`

thomastaylor312 commented:
When you rebase again, if there are still issues, we can take a further look at what is happening

peay commented:
@thomastaylor312 I have rebased again, seems like everything is in order now. Thanks!

thomastaylor312 commented:
Thank you @peay!
